### DISCUSSION TITLE
Access Denied When Executing "GetCharacteristicsForUuidAsync()" for Write Characteristic

### DISCUSSION THREAD

**Discussion by MathiasHartner** [shinyorg/shiny#961] [Q&A] [answered]
Hey guyz,

maybe someone can help me out because I have a really tricky situation with Bluetooth LE using Shiny on Windows 10.

I am using an adapted version of the UWP implementation from Mr. Ritchie. Why that?
Because I need BLE within a Win32 classic desktop application.
Our code is running in a 32 Bit frame application using the .NET runtime (v4.0.30319, .NET Framework 4.6.2). I was able to manage all the other issues (strong naming the Shiny assemblies, running some code in a 64 Bit COM surrogate DLL when running on 64 Bit Windows) but now I am totally stuck with this and here's where:

The problem occurrs within [Uwp/GattService.cs](https://github.com/shinyorg/shiny/blob/master/src/Shiny.BluetoothLE/Platforms/Uwp/GattService.cs) at `var result = await this.native.GetCharacteristicsForUuidAsync(uuid, BluetoothCacheMode.Cached);`.
When getting the read characteristics, the result is `GattCommunicationStatus.Success` and contains the desired characteristic.
But of course I also need to get the write characteristic and I am ALWAYS getting `GattCommunicationStatus.AccessDenied`!
Because of that the code always throws the exception `throw new ArgumentException("GATT Communication failure - " + result.Status);.`

Has anybody a clue why is that? Maybe you @aritchie ? I really need help here because I am kinda lost right now...

I also tried to set `AccessPermission` via registry like supposed [here](https://social.msdn.microsoft.com/Forums/en-US/58da3fdb-a0e1-4161-8af3-778b6839f4e1/bluetooth-bluetoothledevicefromidasync-does-not-complete-on-10015063?forum=wdk) but no luck at all...

P.S. I use Windows 10 SDK Kit Build 20348 and like stated above it is a .NET Framework 4.6.2 project and all our assemblies are strong named because of using GAC. If I am missing anything don't hesitate to contact me.

**Comment by aritchie** [comment]
Hi Mathias

Please respect the guidelines of not tagging maintainers in these posts (https://github.com/shinyorg/shiny/discussions/600).  Please note that I'm only supporting UWP via sponsors at this point since UWP is on it's way out.

I can tell you that I have no idea what to tell you here as it is well outside of I would be willing to support anyhow.  I don't have this issue on pure UWP, so it could be something with the way this has been bridged in... I have no idea.

**Reply by MathiasHartner** [reply]
OK, I'm sorry for tagging you...Did not know that. Sad to hear that you have no idea but anyways thanks for your fast reply.
Maybe there are some other dudes outside who have an idea how to get it to work...

**Comment by MathiasHartner** [comment, marked as answer]
For anyone who stumbles over the same stupid issue...Here's the solution:

On Windows, using .NET Framework 4.6.1+ and the WinRT libraries inside a Non-UWP application, you can only call **ONCE** for getting the characteristics, no matter if you call for all at once via `gattDeviceService.GetCharacteristicsAsync(BluetoothCacheMode.Uncached)` or for a specific one via `gattDeviceService.GetCharacteristicsForUuidAsync(uuid, BluetoothCacheMode.Cached)`.

Any subsequent calls will fail with `GattCommunicationStatus.AccessDenied`...
So my solution now is to retrieve all characteristics at once and filter them locally.

That did the trick! Anyways, this is so stupid...It wasted a lot of my time now!
As it seems, I also do not need to set any `AccessPermission` via registry.

PS: I will call out to you, if I stumble over another tricky situation, just to let you guyz know.
